db/dcrsqlite/sqlite.go

```diff
@@ -912,6 +1037,99 @@ func (db *DB) RetrieveStakeInfoExtended(ind int64) (*apitypes.StakeInfoExtended,
 	return si, nil
 }
 
+func (db *DB) RetrieveStakeInfoExtendedByHash(blockhash string) (*apitypes.StakeInfoExtended, error) {
+	si := apitypes.NewStakeInfoExtended()
+
+	var winners string
+	err := db.QueryRow(db.getStakeInfoExtendedByHashSQL, blockhash).Scan(&si.Hash, &si.Feeinfo.Height,
+		&si.Feeinfo.Number, &si.Feeinfo.Min, &si.Feeinfo.Max, &si.Feeinfo.Mean,
+		&si.Feeinfo.Median, &si.Feeinfo.StdDev,
+		&si.StakeDiff, // no next or estimates
+		&si.PriceWindowNum, &si.IdxBlockInWindow, &si.PoolInfo.Size,
+		&si.PoolInfo.Value, &si.PoolInfo.ValAvg, &winners)
+	if err != nil {
+		return nil, err
+	}
+	si.PoolInfo.Winners = splitToArray(winners)
+	return si, nil
+}
+
+// JustifyTableStructure updates an old structure that wasn't indexing
+// sidechains. It could and should be removed in a future version. The block
+// summary table got two new boolean columns, `is_mainchain` and `is_valid`,
+// and the Primary key was changed from height to hash.
+// The stake info table got a `hash` column and the primary key was switched from height to hash
+func (db *DB) JustifyTableStructures() error {
+
+	// Grab the column info. Right now, just counting them, but could be looked at more closely
+	// Each tuple is a row with columns [cid, name, type, notnull, dflt_value, pk]
+	rows, err := db.Query(db.getBlockSummaryTableInfo)
+	if err != nil {
+		log.Errorf("Query failed: %v", err)
+		return err
+	}
+	defer rows.Close()
+	rowCounter := 0
+	for rows.Next() {
+		rowCounter += 1
+	}
+
+	// It simply checks whether there are enough columns for the current structure
+	if rowCounter >= 12 {
+		return nil
+	}
+
+	log.Info("Detected old SQLite table structure. Updating.")
```

Comment:
It occurred to me that we should really backup the sqlt.db file first.

Reply:
Should I delete the backup if the swap is successful? All table mods are wrapped in a transaction right now, so a failed table copy should be rolled back.

Reply:
It's not a corrupted db I'm concerned with (as you said, failures are rolled back), but devs like us losing our DBs while testing and such.  Although there may be a need to revert to an old version of dcrdata which won't support the new DB format. So I would suggest keeping the old DB copy around regardless of the the migration result.